### <b>Recherche du meilleur combo Mario Kart par la résolution d'un problème d'Optimisation Multi-objectif</b>
<i>par Erwan Martin</i>

Mario kart est une série de jeux vidéo de course éditée par Nintendo depuis 1992. <br>
Le concept est simple, vous incarnez un des personnages de l'univers mario dans une course de karting prenant place sur des pistes de l'univers Mario.
<div style="text-align:center">
    <img src="./img/img2.jpg" alt="Texte alternatif" width="400" height="200">
</div>

Chaque jeu mario kart a sa propre identité et des fonctionnalités qui témoignent des réalités de l'époque de sa sortie. Ils ont chacun fait évoluer le gameplay à leur manière et aujoud'hui, on va surtout s'intéresser à l'une de ces évolutions : La customisation du véhicule.
<div style="text-align:center">
    <img src="./img/img1.jpg" alt="Texte alternatif" width="500" height="300">
</div>

Avant la sortie de Mario Kart 7 en 2011, les performances de votre véhicule étaient largement influencées par le personnage que vous choisissiez. 
<div style="text-align:center">
    <img src="./img/img3.png" alt="Texte alternatif" width="500" height="300">
</div>
Par exemple, si vous incarniez Toad, votre kart aurait une vitesse de pointe relativement faible, mais une excellente maniabilité. En revanche, si vous optiez pour Bowser, les rôles seraient inversés, avec une vitesse de pointe plus élevée mais une maniabilité moindre.

Avant mario kart 7, vous n'aviez donc qu'une variable qui influencait sur vos performances : le <b>choix du personnage</b> (et du <b>chassis</b> dans mario kart DS).<br>

Mario kart 7 va pousser beaucoup plus loin cet aspect du gameplay en introduisant la customisation des <b>roues</b>, du  <b>chassis</b> et du fraichement introduit  <b>deltaplane</b>.
<div style="text-align:center">
    <img src="./img/img4.png" alt="Texte alternatif" width="500" height="300">
</div>

L'abondance de possibilités de personnalisation nous mène naturellement à nous poser la question suivante : <b>Quelle est la combinaison la plus optimale ?</b>

Pour répondre à cette question, nous allons traiter ce problème comme étant un problème d'<b>optimisation multi-objectif</b>. Nous nous concentrerons sur le dernier jeu sortie en date <i><b>Mario Kart 8 Deluxe</b></i> sur Nintendo Switch.

La première étape va être de récupérer toutes les statistiques liées aux éléments customisables. Ces données ne sont pas toutes directement disponible en jeu mais grâce au travail de la communauté mario kart, de nombreux sites spécialisés mettent à dispositions ces statistiques. Il m'a alors suffit de les récupérer manuellement et d'en faire des fichiers csv. 

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
df_perso = pd.read_csv("data/perso.csv",index_col=0)
df_kart = pd.read_csv("data/kart.csv",index_col=0)
df_roues = pd.read_csv("data/roues.csv",index_col=0)
df_delta = pd.read_csv("data/delta.csv",index_col=0)

In [3]:
df_perso

Baby Peach   Baby Daisy  Baby Rosalina   Lemmy Koopa  \
Vitesse             2.50         2.50           2.50          2.50   
Acceleration        4.00         4.00           4.25          4.25   
Poids               2.00         2.00           2.00          2.00   
Maniabilité         5.00         5.00           4.75          4.75   
Adhérence           4.25         4.25           3.75          3.75   
Mini_Turbo          4.00         4.00           4.00          4.00   
Total              21.75        21.75          21.25         21.25   

              Baby Mario/Luigi   Dry Bones  Koopa Troopa   Lakitu  \
Vitesse                   2.75        2.75          3.00     3.00   
Acceleration              4.25        4.25          4.00     4.00   
Poids                     2.25        2.25          2.50     2.50   
Maniabilité               4.50        4.50          4.50     4.50   
Adhérence                 4.00        4.00          4.25     4.25   
Mini_Turbo                4.00        4.00          3.75     3.75   
Total                    21.75       21.75         22.00    22.00   

               Bowser Jr.  Toadette  ...   Pink Gold Peach   Petey Piranha  \
Vitesse              3.00      3.00  ...              4.25            4.25   
Acceleration         4.00      4.25  ...              3.25            3.25   
Poids                2.50      2.50  ...              4.50            4.50   
Maniabilité          4.50      4.25  ...              3.25            3.25   
Adhérence            4.25      3.50  ...              3.25            3.25   
Mini_Turbo           3.75      3.75  ...              3.00            3.00   
Total               22.00     21.25  ...             21.50           21.50   

              Donkey Kong    Roy   Waluigi  Wiggler  Wario   Dry Bowser  \
Vitesse              4.50   4.50      4.50     4.50   4.75         4.75   
Acceleration         3.25   3.25      3.25     3.25   3.00         3.00   
Poids                4.00   4.00      4.00     4.00   4.25         4.25   
Maniabilité          3.00   3.00      3.00     3.00   2.75         2.75   
Adhérence            3.00   3.00      3.00     3.00   3.25         3.25   
Mini_Turbo           3.00   3.00      3.00     3.00   2.75         2.75   
Total               20.75  20.75     20.75    20.75  20.75        20.75   

              Bowser   Morton  
Vitesse         4.75     4.75  
Acceleration    3.00     3.00  
Poids           4.50     4.50  
Maniabilité     2.50     2.50  
Adhérence       3.00     3.00  
Mini_Turbo      2.75     2.75  
Total          20.50    20.50  

[7 rows x 44 columns]

In [4]:
df_roues

Standard  Monster  Roller  Slim  Slick  Metal  Button  Off-Road  \
Vitesse            0.0     0.00   -0.50  0.25   0.50   0.50   -0.25      0.25   
Acceleration       0.0    -0.50    0.50 -0.50  -0.75  -1.00    0.25     -0.25   
Poids              0.0     0.50   -0.50  0.00   0.25   0.50   -0.50      0.25   
Maniabilité        0.0    -0.75    0.25  0.25  -0.25  -0.25    0.00     -0.50   
Adhérence          0.0     0.50   -0.25 -1.00  -1.25  -0.75    0.00      0.25   
Mini_Turbo         0.0    -0.25    0.75 -0.25  -0.75   0.00    0.50     -0.50   
Total              0.0    -0.50    0.25 -1.25  -2.25  -1.00    0.00     -0.50   

              Sponge  Wood  ...  Blue Standard  Hot Monster  Azure Roller  \
Vitesse        -0.25  0.25  ...            0.0         0.00         -0.50   
Acceleration    0.00 -0.50  ...            0.0        -0.50          0.50   
Poids          -0.25  0.00  ...            0.0         0.50         -0.50   
Maniabilité    -0.25  0.25  ...            0.0        -0.75          0.25   
Adhérence       0.25 -1.00  ...            0.0         0.50         -0.25   
Mini_Turbo      0.25 -0.25  ...            0.0        -0.25          0.75   
Total          -0.25 -1.25  ...            0.0        -0.50          0.25   

              Crimson Slim  Cyber Slick  Retro Off-Road  Gold Tires  \
Vitesse               0.25         0.50            0.25        0.50   
Acceleration         -0.50        -0.75           -0.25       -1.00   
Poids                 0.00         0.25            0.25        0.50   
Maniabilité           0.25        -0.25           -0.50       -0.25   
Adhérence            -1.00        -1.25            0.25       -0.75   
Mini_Turbo            0.00        -0.75           -0.50       -0.75   
Total                -1.00        -2.25           -0.50       -1.75   

              GLA Tires  Triforce Tires  Leaf Tires  
Vitesse             0.0            0.25       -0.25  
Acceleration        0.0           -0.25        0.25  
Poids               0.0            0.25       -0.50  
Maniabilité         0.0           -0.50        0.00  
Adhérence           0.0            0.25       -0.50  
Mini_Turbo          0.0           -0.50        0.50  
Total               0.0           -0.50       -0.50  

[7 rows x 21 columns]

In [5]:
df_kart

Standard Kart  Pipe Frame  Mach 8  Cat Cruiser  Steel Driver  \
Vitesse                   0       -0.50    0.00        -0.25          0.25   
Acceleration              0        0.50   -0.25         0.25         -0.75   
Poids                     0       -0.25    0.25         0.00          0.50   
Maniabilité               0        0.50   -0.25         0.25         -0.50   
Adhérence                 0        0.25    0.25         0.00          0.00   
Mini_Turbo                0        0.50    0.00         0.25         -0.50   
Total                     0        1.00    0.00         0.50         -1.00   

              Circuit Special  Tri-Speeder  Badwagon  Prancer  Biddybuggy  \
Vitesse                  0.50         0.25      0.50     0.25       -0.75   
Acceleration            -0.75        -0.75     -1.00    -0.50        0.75   
Poids                    0.25         0.50      0.50    -0.25       -0.50   
Maniabilité             -0.50        -0.50     -0.75     0.00        0.50   
Adhérence               -0.50         0.00      0.50    -0.25        0.50   
Mini_Turbo              -0.75        -0.50     -1.00    -0.25        0.25   
Total                   -1.75        -1.00     -1.25    -1.00        0.75   

              ...  Jet Bike  Yoshi Bike  Master Cycle  City Tripper  \
Vitesse       ...      0.25       -0.25          0.25         -0.50   
Acceleration  ...     -0.50        0.25         -0.50          0.50   
Poids         ...     -0.25        0.00          0.00         -0.25   
Maniabilité   ...      0.00        0.25          0.00          0.50   
Adhérence     ...     -0.25        0.00         -0.75          0.25   
Mini_Turbo    ...     -0.25        0.25         -0.25          0.50   
Total         ...     -1.00        0.50         -1.25          1.00   

              Standard ATV  Wild Wiggler  Teddy Buggy  Bone Rattler  \
Vitesse               0.50         -0.25        -0.25          0.25   
Acceleration         -1.00          0.25         0.25         -0.75   
Poids                 0.50         -0.25         0.00          0.50   
Maniabilité          -0.75          0.25         0.25         -0.50   
Adhérence             0.50          0.50         0.00          0.00   
Mini_Turbo           -1.00          0.25         0.25         -0.50   
Total                -1.25          0.75         0.50         -1.00   

              Inkstriker  Splat Buggy  
Vitesse             0.00         0.25  
Acceleration       -0.25        -0.25  
Poids               0.25        -0.50  
Maniabilité        -0.25        -0.25  
Adhérence           0.25         0.00  
Mini_Turbo          0.00        -0.25  
Total               0.00        -1.00  

[7 rows x 40 columns]

In [6]:
df_delta

Super Glider  Cloud Glider  Wario Wing  Waddle Wing  \
Vitesse                0.0         -0.25        0.00          0.0   
Acceleration           0.0          0.25        0.00          0.0   
Poids                  0.0         -0.25        0.25          0.0   
Maniabilité            0.0          0.00        0.00          0.0   
Adhérence              0.0          0.00       -0.25          0.0   
Mini_Turbo             0.0          0.25        0.00          0.0   
Total                  0.0          0.00        0.00          0.0   

              Peach Parasol  Parachute  Parafoil  Flower Glider  Bowser Kite  \
Vitesse               -0.25      -0.25     -0.25          -0.25        -0.25   
Acceleration           0.25       0.25      0.25           0.25         0.25   
Poids                 -0.25      -0.25      0.00          -0.25         0.00   
Maniabilité            0.00       0.00      0.00           0.00         0.00   
Adhérence              0.00       0.00     -0.25           0.00        -0.25   
Mini_Turbo             0.25       0.25      0.25           0.25         0.25   
Total                  0.00       0.00      0.00           0.00         0.00   

              Plane Glider  MKTV Parafoil  Gold Glider  Hylian Kite  \
Vitesse               0.00          -0.25         0.00          0.0   
Acceleration          0.00           0.25         0.00          0.0   
Poids                 0.25           0.00         0.25          0.0   
Maniabilité           0.00           0.00         0.00          0.0   
Adhérence            -0.25          -0.25        -0.25          0.0   
Mini_Turbo            0.00           0.25         0.00          0.0   
Total                 0.00           0.00         0.00          0.0   

              Paper Glider  
Vitesse              -0.25  
Acceleration          0.25  
Poids                -0.25  
Maniabilité           0.00  
Adhérence             0.00  
Mini_Turbo            0.25  
Total                 0.00

In [7]:
print(df_perso.shape)
print(df_roues.shape)
print(df_kart.shape)
print(df_delta.shape)

(7, 44)
(7, 21)
(7, 40)
(7, 14)


Nous disposons de 44 personnages, 21 roues, 40 chassis et 14 deltaplane. <br>
Nous avons donc 44 x 21 x 40 x 14 = 517 440 combinaisons possibles.<br>
Étant donné que le nombre de possibilité est fini, on peut générer un ensemble d'observations comportant toutes les observations possibles.

On définie une observation comme étant un vecteur de dimension 6 qui pour une combinaison donnée comporte :
- La vitesse totale : La vitesse de pointe de notre véhicule.
- L'accelération totale : L'acceleration de notre véhicule (capacité à changer de vitesse +/- rapidement)
- Le poids total : Le poids du véhicule (utile lors de coalision avec les autres joueurs)
- La maniabilité totale : La maniabilité du véhicule.
- L'adhérence totale : L'adhérence du véhicule sur des terrains pluvieux, sablonneux ou enneigés.
- Le mini turbo totale : La capacité à sortir un mini turbo* +/- rapidement après un dérapage.

<u>Mini turbo</u> : Dans Mario kart, lorsque vous effectuez un dérapage (notamment durant un virage) et que vous le maintenez assez longtemps le jeu vous donne un bonus de vitesse que l'on appelle 'Mini turbo'.
<div style="text-align:center">
    <img src="./img/img6.jpg" alt="Texte alternatif" width="500" height="300">
</div>

In [8]:
#Génération d'une observation (= d'un individu)
def generate_indiv(perso,roue,kart,delta,df_perso,df_roues,df_kart,df_delta):
    Xi = {}
    Xi["Combos"] = "{},{},{},{}".format(perso,roue,kart,delta)
    
    Xi["Vitesse"] = float(df_perso[perso][0])+float(df_delta[delta][0])+float(df_kart[kart][0])+float(df_roues[roue][0])
    Xi["Acceleration"] = float(df_perso[perso][1])+float(df_delta[delta][1])+float(df_kart[kart][1])+float(df_roues[roue][1])
    Xi["Poids"] = float(df_perso[perso][2])+float(df_delta[delta][2])+float(df_kart[kart][2])+float(df_roues[roue][2])
    Xi["Maniabilité"] = float(df_perso[perso][3])+float(df_delta[delta][3])+float(df_kart[kart][3])+float(df_roues[roue][3])
    Xi["Adhérence"] = float(df_perso[perso][4])+float(df_delta[delta][4])+float(df_kart[kart][4])+float(df_roues[roue][4])
    Xi["Mini_Turbo"] = float(df_perso[perso][5])+float(df_delta[delta][5])+float(df_kart[kart][5])+float(df_roues[roue][5])
    return Xi 

#Génération d'une population (= d'un ensemble d'individus)
def generate_pop(df_perso,df_roues,df_kart,df_delta):

    personnages = df_perso.columns
    roues = df_roues.columns
    karts = df_kart.columns
    deltas = df_delta.columns
    X = [] #Liste qui contiendra toutes les combinaisons possibles
    #Pour chaque combinaison (perso,roue,kart,delta) possible, X apprend le vecteur Xi correspondant
    for perso in personnages:
        for roue in roues:
            for kart in karts:
                for delta in deltas:
                    Xi = generate_indiv(perso,roue,kart,delta,df_perso,df_roues,df_kart,df_delta)
                    X.append(Xi)
                    
    return X
    

In [9]:
X = generate_pop(df_perso,df_roues,df_kart,df_delta)

In [10]:
DF = pd.DataFrame(X)

In [11]:
DF

Combos  Vitesse  \
0        Baby Peach,Standard,Standard Kart,Super Glider     2.50   
1        Baby Peach,Standard,Standard Kart,Cloud Glider     2.25   
2          Baby Peach,Standard,Standard Kart,Wario Wing     2.50   
3         Baby Peach,Standard,Standard Kart,Waddle Wing     2.50   
4       Baby Peach,Standard,Standard Kart,Peach Parasol     2.25   
...                                                 ...      ...   
517435       Morton,Leaf Tires,Splat Buggy,Plane Glider     4.75   
517436      Morton,Leaf Tires,Splat Buggy,MKTV Parafoil     4.50   
517437        Morton,Leaf Tires,Splat Buggy,Gold Glider     4.75   
517438        Morton,Leaf Tires,Splat Buggy,Hylian Kite     4.75   
517439       Morton,Leaf Tires,Splat Buggy,Paper Glider     4.50   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  
0               4.00   2.00         5.00       4.25        4.00  
1               4.25   1.75         5.00       4.25        4.25  
2               4.00   2.25         5.00       4.00        4.00  
3               4.00   2.00         5.00       4.25        4.00  
4               4.25   1.75         5.00       4.25        4.25  
...              ...    ...          ...        ...         ...  
517435          3.00   3.75         2.25       2.25        3.00  
517436          3.25   3.50         2.25       2.25        3.25  
517437          3.00   3.75         2.25       2.25        3.00  
517438          3.00   3.50         2.25       2.50        3.00  
517439          3.25   3.25         2.25       2.50        3.25  

[517440 rows x 7 columns]

On a bien un dataframe de 517 440 lignes tel que pour chaque ligne nous avons une colonne de description du vecteur et 6 colonnes qui correspondent aux éléments de ce vecteur.<br>

Nous disposons à présent de notre population, il nous faut maintenant déterminer les fonctions objectifs.<br>
Ce sont les fonctions qui détermineront si une solution au problème est meilleur qu'une autre.

Nous disposons de 6 variables, mais toutes ne sont pas intéressantes à optimiser pour répondre à notre problème.
En effet, on cherche à trouver la combinaison qui nous permettrait de gagner le plus de courses possible. Pour rappel, nous disposons de ces 6 variables :<br>
Le personnage, la vitesse, l'accélération, le poids, l'adhérence, la maniabilité et le mini turbo (variable cachée en jeu).<br>
<div style="text-align:center">
    <img src="./img/img5.jpg" alt="Texte alternatif" width="500" height="300">
</div>

On peut déjà écarter la variable 'Poids' car le poids en lui même n'a que très peu d'influence sur le gameplay.<br>

On va ensuite écarter les variables Adhérence et Maniabilité car bien qu'il soit agréable d'avoir plus de contrôle sur son véhicule, ce sont des variables qui peuvent être compensées par vos compétences en tant que joueur.<br>

En effet, un joueur expérimenté sera capable de manier son véhicule parfaitement quelque soit le terrain ou le véhicule. Ainsi dans cette optique, il est préférable d'essayer d'optimiser d'autres paramètres en priorité.<br>

Il nous reste donc la vitesse, le mini turbo et l'accélération. Nous allons faire le choix sujet à débat d'écarter l'accélération. L'accélération dans mario kart est surtout utile pour redémarrer rapidement après avoir reçu un objet de la part des adversaires. Mais si on part du postulat qu'un joueur expérimenté saura faire en sorte de recevoir le moins d'objets possible, alors la statistique d'accélération devient moins importante. <br>

C'est un choix discutable, car il existe des situations où même en étant le meilleur joueur du monde, on ne pourra pas faire autrement que de se prendre des objets. Mais est-ce que ces situations suffisent à justifier d'empiéter sur l'optimisation de la vitesse et du mini turbo pour améliorer l'accélération ? Dans cette étude, j'ai décidé que non.

Nous avons alors deux fonctions objectifs :
- f1 : Décrit la vitesse de pointe du véhicule en fonction de la combinaison choisie
- f2 : Décrit le mini turbo total du véhicule en fonction de la combinaison choisie 

On va chercher à maximiser ces deux fonctions objectifs en explorant la population dont nous disposons

In [ ]:
fig = px.scatter(DF, x="Vitesse", y="Mini_Turbo",hover_name="Combos",title="Vitesse et Mini Turbo de l'ensemble des combinaisons possibles")
fig.show()

<div style="text-align:center">
    <img src="./img/img10.png" alt="Texte alternatif" width="500" height="300">
</div>

On va à présent s'intérésser aux points qui se situent sur la <b>frontière de Pareto</b>. Nous retrouvons sur cette frontière l'ensemble des optimas de Pareto. C'est-à-dire que sur cette frontière, il n'existe objectivement aucune solution meilleur que l'autre. On parle de solutions non dominées.<br>

<u>Exemple</u> : une solution s1 offre 5.5 en vitesse et 2.75 en mini turbo, une solution s2 offre 5.75 en vitesse et 2.25 en mini turbo. <br>

s1 offre un meilleur mini turbo que s2 mais s2 offre une meilleure vitesse que s1. Comment choisir?<br> 

Il faudrait faire une étude plus poussée en jeu pour savoir qu'est-ce qui est le plus important entre la vitesse le mini turbo et à quel point.<br> 

Pour l'instant, ce que je peux dire, d'après mes recherches, c'est que cela va grandement dépendre du choix de la piste. Sur une piste avec beaucoup de virages, il sera préférable de favoriser la statistique de mini turbo tandis que sur une piste avec beaucoup de ligne droite, la vitesse de pointe sera plus importante.

In [35]:
DF

Combos  Vitesse  \
0        Baby Peach,Standard,Standard Kart,Super Glider     2.50   
1        Baby Peach,Standard,Standard Kart,Cloud Glider     2.25   
2          Baby Peach,Standard,Standard Kart,Wario Wing     2.50   
3         Baby Peach,Standard,Standard Kart,Waddle Wing     2.50   
4       Baby Peach,Standard,Standard Kart,Peach Parasol     2.25   
...                                                 ...      ...   
517435       Morton,Leaf Tires,Splat Buggy,Plane Glider     4.75   
517436      Morton,Leaf Tires,Splat Buggy,MKTV Parafoil     4.50   
517437        Morton,Leaf Tires,Splat Buggy,Gold Glider     4.75   
517438        Morton,Leaf Tires,Splat Buggy,Hylian Kite     4.75   
517439       Morton,Leaf Tires,Splat Buggy,Paper Glider     4.50   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  
0               4.00   2.00         5.00       4.25        4.00  
1               4.25   1.75         5.00       4.25        4.25  
2               4.00   2.25         5.00       4.00        4.00  
3               4.00   2.00         5.00       4.25        4.00  
4               4.25   1.75         5.00       4.25        4.25  
...              ...    ...          ...        ...         ...  
517435          3.00   3.75         2.25       2.25        3.00  
517436          3.25   3.50         2.25       2.25        3.25  
517437          3.00   3.75         2.25       2.25        3.00  
517438          3.00   3.50         2.25       2.50        3.00  
517439          3.25   3.25         2.25       2.50        3.25  

[517440 rows x 7 columns]

In [39]:
DF.sort_values(by='Vitesse').index[0]

1259

In [81]:
def pareto_frontier(df, objective1, objective2, label_column):
    # On trie le dataframe en fonction du 1er objectif (ici la vitesse)
    sorted_df = df.sort_values(by=objective1)

    # On initialise la frontière de Pareto avec le premier point (qui n'est pas un point dominé grâce au tri)
    frontier = [(sorted_df.iloc[0][objective1], sorted_df.iloc[0][objective2], sorted_df.iloc[0][label_column],sorted_df.index[0])]
    # On regarde pour chaque point s'il est dominé ou non
    for index, row in sorted_df.iterrows():
        current_point = (row[objective1], row[objective2], row[label_column],index)
        is_dominated = False

        for point in frontier:
            # On vérifie si le point actuel domine l'un des points de la frontière de Pareto actuelle
            if (current_point[0] >= point[0]) and (current_point[1] >= point[1]):
                # Si c'est le cas, on retire le point de la frontière
                frontier.remove(point)

            # On vérifie si le point actuel est dominé par l'un des points de la frontière
            elif (current_point[0] <= point[0]) and (current_point[1] <= point[1]):
                #Si c'est le cas, on le tag comme étant dominé
                is_dominated = True
                break

        # S'il n'est pas dominé, on l'ajoute à la frontière
        if not is_dominated:
            frontier.append(current_point)

    return frontier




In [82]:
frontier_points = pareto_frontier(DF, 'Vitesse', 'Mini_Turbo', 'Combos')
frontier_points

[(1.75, 6.0, ' Dry Bones,Azure Roller,Tanooki Kart,Flower Glider', 66339),
 (1.75, 6.0, ' Dry Bones,Azure Roller,Tanooki Kart,Parafoil', 66338),
 (2.75, 5.75, ' Villager Boy,Azure Roller,Tanooki Kart,Paper Glider', 289785),
 (2.75, 5.75, ' Villager Boy,Azure Roller,Koopa Clown,MKTV Parafoil', 289838),
 (3.0, 5.5, ' Daisy,Azure Roller,Tanooki Kart,Super Glider', 230972),
 (3.25, 5.25, ' Kamek,Leaf Tires,Koopa Clown,Cloud Glider', 329029),
 (3.75, 5.0, ' Inkling Boy,Metal,Koopa Clown,Paper Glider', 273601),
 (3.75, 5.0, ' Inkling Boy,Metal,Tanooki Kart,Cloud Glider', 273533),
 (4.0, 4.75, ' Iggy,Metal,Tanooki Kart,Paper Glider', 308825),
 (4.25, 4.5, ' Iggy,Metal,Tanooki Kart,Hylian Kite', 308824),
 (4.5, 4.25, ' Roy,Metal,Koopa Clown,Paper Glider', 438241),
 (4.75, 4.0, ' Roy,Metal,Koopa Clown,Waddle Wing', 438231),
 (5.0, 3.75, 'Wario,Metal,Tanooki Kart,Waddle Wing', 473455),
 (5.25, 3.0, 'Bowser,Metal,Blue Falcon,Paper Glider', 496971),
 (5.5, 2.75, ' Morton,Metal,Blue Falcon,Wario Wi

In [73]:
indexes = [item[-1] for item in frontier_points]

In [90]:
DF['Pareto'] = 0
DF.loc[indexes,'Pareto'] = 1

In [78]:
DF[DF['Pareto']==1]

Combos  Vitesse  \
66338        Dry Bones,Azure Roller,Tanooki Kart,Parafoil     1.75   
66339    Dry Bones,Azure Roller,Tanooki Kart,Flower Gl...     1.75   
230972       Daisy,Azure Roller,Tanooki Kart,Super Glider     3.00   
273533        Inkling Boy,Metal,Tanooki Kart,Cloud Glider     3.75   
273601         Inkling Boy,Metal,Koopa Clown,Paper Glider     3.75   
289785   Villager Boy,Azure Roller,Tanooki Kart,Paper ...     2.75   
289838   Villager Boy,Azure Roller,Koopa Clown,MKTV Pa...     2.75   
308824                Iggy,Metal,Tanooki Kart,Hylian Kite     4.25   
308825               Iggy,Metal,Tanooki Kart,Paper Glider     4.00   
329029          Kamek,Leaf Tires,Koopa Clown,Cloud Glider     3.25   
438231                  Roy,Metal,Koopa Clown,Waddle Wing     4.75   
438241                 Roy,Metal,Koopa Clown,Paper Glider     4.50   
473455               Wario,Metal,Tanooki Kart,Waddle Wing     5.00   
496971              Bowser,Metal,Blue Falcon,Paper Glider     5.25   
496998                  Bowser,Metal,B Dasher,Hylian Kite     5.75   
508720                Morton,Metal,Blue Falcon,Wario Wing     5.50   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  Pareto  
66338           4.50   2.00         5.00       3.50        6.00       1  
66339           4.50   1.75         5.00       3.75        6.00       1  
230972          3.75   2.75         4.25       3.50        5.50       1  
273533          2.50   3.75         3.75       2.50        5.00       1  
273601          2.50   3.75         3.75       2.50        5.00       1  
289785          4.00   2.75         4.25       3.00        5.75       1  
289838          4.00   3.00         4.25       2.75        5.75       1  
308824          2.00   4.25         3.75       2.50        4.50       1  
308825          2.25   4.00         3.75       2.50        4.75       1  
329029          3.50   3.00         4.00       2.75        5.25       1  
438231          1.75   4.75         3.00       2.25        4.00       1  
438241          2.00   4.50         3.00       2.25        4.25       1  
473455          1.50   5.00         2.75       2.50        3.75       1  
496971          2.00   4.25         2.00       2.75        3.00       1  
496998          1.25   5.25         1.75       2.00        2.25       1  
508720          1.75   4.75         2.00       2.50        2.75       1

In [92]:
for index in indexes:
    speed = DF.loc[index,'Vitesse']
    turbo = DF.loc[index,'Mini_Turbo']
    DF.loc[(DF['Vitesse']==speed) & (DF['Mini_Turbo']==turbo),'Pareto']=1

In [ ]:
fig = px.scatter(DF, x="Vitesse", y="Mini_Turbo",hover_name="Combos",color='Pareto',color_continuous_scale=['blue', 'red'],title="Vitesse et Mini Turbo de l'ensemble des combinaisons possibles")
fig.show()

<div style="text-align:center">
    <img src="./img/img11.png" alt="Texte alternatif" width="500" height="300">
</div>

In [112]:
DF['Pareto'].value_counts()


Pareto
0    515516
1      1924
Name: count, dtype: int64

On a donc éliminé 99.6% des combos possibles en jeu, il nous reste alors 1924 combos viables.

On a également des combos qui donnent exactement les mêmes statistiques et diffèrent uniquement par le design, si on enlève les doublons on se retrouve avec une liste de 66 combos viables.

In [116]:
pareto = DF[DF['Pareto']==1]

In [141]:
pareto_reduced = pareto.drop_duplicates(subset=pareto.drop(columns=['Combos']).columns)

In [142]:
pareto_reduced

Combos  Vitesse  \
48413   Baby Mario/Luigi,Roller,Tanooki Kart,Cloud Glider     1.75   
48418       Baby Mario/Luigi,Roller,Tanooki Kart,Parafoil     1.75   
213052             Peach,Roller,Tanooki Kart,Super Glider     3.00   
213053             Peach,Roller,Tanooki Kart,Cloud Glider     2.75   
213054               Peach,Roller,Tanooki Kart,Wario Wing     3.00   
...                                                   ...      ...   
496973             Bowser,Metal,Tanooki Kart,Cloud Glider     4.75   
496974               Bowser,Metal,Tanooki Kart,Wario Wing     5.00   
496978                 Bowser,Metal,Tanooki Kart,Parafoil     4.75   
496986                 Bowser,Metal,B Dasher,Super Glider     5.75   
496988                   Bowser,Metal,B Dasher,Wario Wing     5.75   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  Pareto  
48413           4.50   1.75         5.00       3.75        6.00       1  
48418           4.50   2.00         5.00       3.50        6.00       1  
213052          3.75   2.75         4.25       3.50        5.50       1  
213053          4.00   2.50         4.25       3.50        5.75       1  
213054          3.75   3.00         4.25       3.25        5.50       1  
...              ...    ...          ...        ...         ...     ...  
496973          1.75   5.00         2.50       2.25        4.00       1  
496974          1.50   5.50         2.50       2.00        3.75       1  
496978          1.75   5.25         2.50       2.00        4.00       1  
496986          1.25   5.25         1.75       2.00        2.25       1  
496988          1.25   5.50         1.75       1.75        2.25       1  

[70 rows x 8 columns]

à partir de cet instant, tous les combos ici présent sont viables et tout va dépendre de vos préférences. Cependant, en pratique, on peut constater qu'une tendance générale se dessine en mode online :
<div style="text-align:center">
    <img src="./img/img8.jpg" alt="Texte alternatif" width="500" height="300">
</div>

Il semble que le combo, Tanooki Kart, Roller avec Yoshi (ou Daisy, Peach qui partagent les mêmes statistiques) soit plus populaire, intéressons nous donc à ce combo.

Il se situe effectivement sur notre frontier de pareto, il correspond au point Vitesse 3.0 et Mini Turbo 5.5
<div style="text-align:center">
    <img src="./img/img9.png" alt="Texte alternatif" width="500" height="300">
</div>

C'est un point qui cherche un compromis raisonnable entre la vitesse et le mini turbo, cependant ce n'est pas le seul combo proposant ces statistiques donc intéressons nous autres statistiques que nous avons mis de côté jusque là.

In [148]:
p = pareto_reduced[(pareto_reduced['Vitesse']==3.0) & (pareto_reduced['Mini_Turbo']==5.5)].sort_values('Adhérence')
p

Combos  Vitesse  \
270178     Tanooki Mario,Leaf Tires,Tanooki Kart,Parafoil      3.0   
260094       Tanooki Mario,Roller,Tanooki Kart,Wario Wing      3.0   
270173  Tanooki Mario,Leaf Tires,Tanooki Kart,Cloud Gl...      3.0   
223138             Peach,Leaf Tires,Tanooki Kart,Parafoil      3.0   
260092     Tanooki Mario,Roller,Tanooki Kart,Super Glider      3.0   
262338         Tanooki Mario,Button,Tanooki Kart,Parafoil      3.0   
213054               Peach,Roller,Tanooki Kart,Wario Wing      3.0   
223133         Peach,Leaf Tires,Tanooki Kart,Cloud Glider      3.0   
262333     Tanooki Mario,Button,Tanooki Kart,Cloud Glider      3.0   
213052             Peach,Roller,Tanooki Kart,Super Glider      3.0   
215298                 Peach,Button,Tanooki Kart,Parafoil      3.0   
215293             Peach,Button,Tanooki Kart,Cloud Glider      3.0   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  Pareto  
270178          3.75   3.00         4.00       2.50         5.5       1  
260094          3.75   3.25         4.25       2.75         5.5       1  
270173          3.75   2.75         4.00       2.75         5.5       1  
223138          3.75   2.75         4.00       3.00         5.5       1  
260092          3.75   3.00         4.25       3.00         5.5       1  
262338          3.75   3.00         4.00       3.00         5.5       1  
213054          3.75   3.00         4.25       3.25         5.5       1  
223133          3.75   2.50         4.00       3.25         5.5       1  
262333          3.75   2.75         4.00       3.25         5.5       1  
213052          3.75   2.75         4.25       3.50         5.5       1  
215298          3.75   2.75         4.00       3.50         5.5       1  
215293          3.75   2.50         4.00       3.75         5.5       1

In [149]:
p['sum'] = p['Maniabilité']+p['Adhérence']

In [151]:
p.sort_values('sum',ascending=False)

Combos  Vitesse  \
213052             Peach,Roller,Tanooki Kart,Super Glider      3.0   
215293             Peach,Button,Tanooki Kart,Cloud Glider      3.0   
213054               Peach,Roller,Tanooki Kart,Wario Wing      3.0   
215298                 Peach,Button,Tanooki Kart,Parafoil      3.0   
260092     Tanooki Mario,Roller,Tanooki Kart,Super Glider      3.0   
223133         Peach,Leaf Tires,Tanooki Kart,Cloud Glider      3.0   
262333     Tanooki Mario,Button,Tanooki Kart,Cloud Glider      3.0   
260094       Tanooki Mario,Roller,Tanooki Kart,Wario Wing      3.0   
223138             Peach,Leaf Tires,Tanooki Kart,Parafoil      3.0   
262338         Tanooki Mario,Button,Tanooki Kart,Parafoil      3.0   
270173  Tanooki Mario,Leaf Tires,Tanooki Kart,Cloud Gl...      3.0   
270178     Tanooki Mario,Leaf Tires,Tanooki Kart,Parafoil      3.0   

        Acceleration  Poids  Maniabilité  Adhérence  Mini_Turbo  Pareto   sum  
213052          3.75   2.75         4.25       3.50         5.5       1  7.75  
215293          3.75   2.50         4.00       3.75         5.5       1  7.75  
213054          3.75   3.00         4.25       3.25         5.5       1  7.50  
215298          3.75   2.75         4.00       3.50         5.5       1  7.50  
260092          3.75   3.00         4.25       3.00         5.5       1  7.25  
223133          3.75   2.50         4.00       3.25         5.5       1  7.25  
262333          3.75   2.75         4.00       3.25         5.5       1  7.25  
260094          3.75   3.25         4.25       2.75         5.5       1  7.00  
223138          3.75   2.75         4.00       3.00         5.5       1  7.00  
262338          3.75   3.00         4.00       3.00         5.5       1  7.00  
270173          3.75   2.75         4.00       2.75         5.5       1  6.75  
270178          3.75   3.00         4.00       2.50         5.5       1  6.50

On constate qu'au niveau de l'accélération, il propose tous la même chose mais diffère surtout au niveau de l'adhérence et de la maniabilité très légèrement. <br>

Quand on fait la somme de ces deux caractéristiques, on se rend compte que le Combo Peach (Yoshi, daisy) avec les roues Roller et le chassis Tanooki Kart est un peu meilleur que les autres. C'est ce qui explique pourquoi il est largement utilisé en ligne !<br>